File: 355.design-twitter.py
#
# @lc app=leetcode id=355 lang=python3
#
# [355] Design Twitter
#

# Time:  O(klogu), k is most recently number of tweets,
#                  u is the number of the user's following.
# Space: O(t + f), t is the total number of tweets,
#                  f is the total number of followings.

from collections import defaultdict
from typing import List
import heapq


# @lc code=start
class Twitter:
    def __init__(self):
        self.time = 0
        self.most_recent_n = 10
        self.followings = defaultdict(set)
        self.messages = defaultdict(list)

    def postTweet(self, userId: int, tweetId: int) -> None:
        self.time += 1
        self.messages[userId].append((self.time, tweetId))

    def getNewsFeed(self, userId: int) -> List[int]:
        max_heap = []  # heapq is min_heap so *= -1
        if self.messages[userId]:
            heapq.heappush(max_heap, (-self.messages[userId][-1][0], userId, 0))
        for uid in self.followings[userId]:
            if self.messages[uid]:
                heapq.heappush(max_heap, (-self.messages[uid][-1][0], uid, 0))
        res = []
        while max_heap and len(res) < self.most_recent_n:
            time, uid, curr = heapq.heappop(max_heap)
            next = curr + 1
            if next != len(self.messages[uid]):
                heapq.heappush(
                    max_heap, (-self.messages[uid][-(next + 1)][0], uid, next)
                )
            res.append(self.messages[uid][-(curr + 1)][1])
        return res

    def follow(self, followerId: int, followeeId: int) -> None:
        if followerId != followeeId:
            self.followings[followerId].add(followeeId)

    def unfollow(self, followerId: int, followeeId: int) -> None:
        self.followings[followerId].discard(followeeId)


# Your Twitter object will be instantiated and called as such:
# obj = Twitter()
# obj.postTweet(userId,tweetId)
# param_2 = obj.getNewsFeed(userId)
# obj.follow(followerId,followeeId)
# obj.unfollow(followerId,followeeId)
# @lc code=end

File: 355.design-twitter.js
/*
 * @lc app=leetcode id=355 lang=javascript
 *
 * [355] Design Twitter
 */

// @lc code=start
/**
 * Initialize your data structure here.
 */
var Twitter = function () {
  this.time = 0;
  this.tweets = {};
  this.friends = {};
};

/**
 * Compose a new tweet.
 * @param {number} userId
 * @param {number} tweetId
 * @return {void}
 */
Twitter.prototype.postTweet = function (userId, tweetId) {
  this.follow(userId, userId);
  if (!this.tweets[userId]) {
    this.tweets[userId] = new Map();
  }
  this.tweets[userId].set(this.time++, tweetId);
};

/**
 * Retrieve the 10 most recent tweet ids in the user's news feed. Each item in the news feed must be posted by users who the user followed or by the user herself. Tweets must be ordered from most recent to least recent.
 * @param {number} userId
 * @return {number[]}
 */
Twitter.prototype.getNewsFeed = function (userId) {
  if (!this.friends[userId]) {
    return [];
  }
  const res = [];
  const top10 = new Map();
  for (const id of this.friends[userId].keys()) {
    if (!this.tweets[id]) {
      continue;
    }
    for (const t of this.tweets[id]) {
      top10.set(t[0], t[1]);
      if (top10.size > 10) {
        const earliest = Array.from(top10.keys()).sort(
          (a, b) => Number(a) - Number(b)
        )[0];
        top10.delete(earliest);
      }
    }
  }
  for (const k of Array.from(top10.keys()).sort(
    (a, b) => Number(b) - Number(a)
  )) {
    res.push(top10.get(k));
  }
  return res;
};

/**
 * Follower follows a followee. If the operation is invalid, it should be a no-op.
 * @param {number} followerId
 * @param {number} followeeId
 * @return {void}
 */
Twitter.prototype.follow = function (followerId, followeeId) {
  if (!this.friends[followerId]) {
    this.friends[followerId] = new Set();
  }
  this.friends[followerId].add(followeeId);
};

/**
 * Follower unfollows a followee. If the operation is invalid, it should be a no-op.
 * @param {number} followerId
 * @param {number} followeeId
 * @return {void}
 */
Twitter.prototype.unfollow = function (followerId, followeeId) {
  if (followerId !== followeeId) {
    this.friends[followerId].delete(followeeId);
  }
};

/**
 * Your Twitter object will be instantiated and called as such:
 * var obj = new Twitter()
 * obj.postTweet(userId,tweetId)
 * var param_2 = obj.getNewsFeed(userId)
 * obj.follow(followerId,followeeId)
 * obj.unfollow(followerId,followeeId)
 */
// @lc code=end
